adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/PropertyBag.java

```diff
@@ -0,0 +1,99 @@
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.hono.adapter.mqtt;
+
+import io.netty.handler.codec.http.QueryStringDecoder;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * A collection of methods for processing <em>property-bag</em> set at the end of a topic.
+ *
+ */
+public final class PropertyBag {
+
+    private final Map<String, List<String>> properties;
+    private final String topic;
+    private final String topicWithoutPropertyBag;
+
+    private PropertyBag(final String topic) {
+        this.topic = topic;
+        this.properties = getProperties();
+        this.topicWithoutPropertyBag = extractTopicWithoutPropertyBag();
+    }
+
+    /**
+     * Creates a property bag object from the given topic.
+     * 
+     * @param topic The topic that the message has been published to.
+     * @return The property bag object.
+     * @throws NullPointerException if topic is {@code null}.
+     */
+    public static PropertyBag fromTopic(final String topic) {
+        Objects.requireNonNull(topic);
+        return new PropertyBag(topic);
+    }
+
+    /**
+     * Gets a property value from the <em>property-bag</em>.
+     *
+     * @param name The property name.
+     * @return The property value or {@code null} if the property is not set.
+     */
+    public String getProperty(final String name) {
+        return Optional.ofNullable(properties)
+                .map(props -> props.get(name))
+                .map(values -> values.get(0))
+                .orElse(null);
+    }    
+
+    /**
+     * Returns the topic without the <em>property-bag</em>.
+     *
+     * @return The topic without the <em>property-bag</em>.
+     */
+    public String topicWithoutPropertyBag() {
+        return topicWithoutPropertyBag;
+    }
+
+    /**
+     * Trims the <em>property-bag</em> from the topic and returns the rest.
+     *
+     * @return The topic without the <em>property-bag</em>.
+     */
+    private String extractTopicWithoutPropertyBag() {
+        final int index = topic.lastIndexOf("?");
+        if (index > 0) {
+            return topic.substring(0, index);
+        }
+        return topic;
+    }
+
+    /**
+     * Retrieves the <em>property-bag</em> from the given topic and 
+     * extracts the key/value pairs into a map.
+     * 
+     * @return The properties from the <em>property-bag</em> or {@code null}
+     *         if no <em>property-bag</em> is present or the topic is {@code null}.
+     */
+    private Map<String, List<String>> getProperties() {
+        final int index = topic.lastIndexOf("?");
```

Comment:
IMHO this is not precise enough. A topic name may as well contain segments that contain `?`.
I think that we need to check for the last segment of the topic name to actually begin with the `?` character, right?

Reply:
@sophokles73 Ah I see the point now regarding additional topic segments. A topic such as `event/tenant/device/seg1/seg2?param=1/seg3/seg4` which has no `property-bag ` will be considered as the one with the `property-bag`. I think `topic.lastIndexOf("/?")` instead of `topic.lastIndexOf("?")` should do the trick. The last index of the separator (/) will be used to identify the last segment and the marker (?) to identify the beginning of the `property-bag`. WDYT?

Reply:
I think so, yes. Consequently, I think that the *topic without the property-bag* should not include the trailing `/` ...

Reply:
:+1:

Reply:
@sophokles73 I have replaced it with `topic.lastIndexOf("/?")` with an additional test. Would you mind taking a look.